Kudu.Core/Deployment/StructuredTextLogger.cs

```diff
@@ -0,0 +1,120 @@
+﻿using Kudu.Core.Tracing;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.RegularExpressions;
+
+namespace Kudu.Core.Deployment
+{
+    public class StructuredTextLogger : IDetailedLogger
+    {
+        private readonly IAnalytics _analytics;
+        private readonly int _depth;
+        private readonly string _path;
+        private const string LogEntryTemplate = "message=\"{0}\" time=\"{1}\" id=\"{2}\" type=\"{3}\"";
+        private const string LogEntryPattern = "message=\"(.*?)\"\\s+time=\"(.*?)\"\\s+id=\"(.*?)\"\\s+type=\"(.*?)\"";
+        private readonly Regex LogEntryRegex = new Regex(LogEntryPattern, RegexOptions.IgnoreCase | RegexOptions.Singleline);
+        private readonly StructuredTextDocument<LogEntry> _structuredTextDocment;
+
+        private static IEnumerable<KeyValuePair<string, string>> EscapeChars = new Dictionary<string, string>
+        {
+            { "\"", "&quot;" }
+        }
+        .Union(StructuredTextDocument.NotAllowedSequances);
+
+        public StructuredTextLogger(string path, IAnalytics analytics, int depth = 0)
+        {
+            _depth = depth;
+            _path = path;
+            _analytics = analytics;
+            _structuredTextDocment = new StructuredTextDocument<LogEntry>(path,
+                e => string.Format(LogEntryTemplate, e.Message, e.LogTime, e.Id, (int)e.Type),
+                str =>
+                {
+                    var match = LogEntryRegex.Match(str);
+                    if (match.Success)
+                    {
+                        var message = UnsanatizeValue(match.Groups[1].Value);
+                        var time = DateTime.Parse(match.Groups[2].Value);
+                        var id = match.Groups[3].Value;
+                        var type = (LogEntryType)Int32.Parse(match.Groups[4].Value);
+                        return new LogEntry(time, id, message, type);
+                    }
+                    else
+                    {
+                        throw new FormatException(string.Format("the log line \"{0}\" is in an invalid format", str));
+                    }
+                });
+        }
+
+        public ILogger Log(string value, LogEntryType type)
+        {
+            try
+            {
+                value = SanatizeValue(value);
+                _structuredTextDocment.Write(new LogEntry(DateTime.UtcNow, Guid.NewGuid().ToString(), value, type), _depth);
```

Comment:
it seems that [XDocument pretty much re-implements how DateTime is serialized ](http://referencesource.microsoft.com/#System.Xml/System/Xml/Schema/XsdDateTime.cs,505) Though I think `.ToString("o")` has the same output we want.